csrc/device_lower/analysis/tensor_memory.cpp

```diff
@@ -7,12 +7,59 @@
 // clang-format on
 
 #include <device_lower/analysis/tensor_memory.h>
+#include <device_lower/lower2device.h>
 #include <fusion.h>
 #include <ir/all_nodes.h>
 #include <type.h>
 
+#include <utility>
+#include <vector>
+
 namespace nvfuser {
 
+// Returns the lane and column allocation domain that is actually allocated.
+std::pair<std::vector<IterDomain*>, std::vector<IterDomain*>> getTMemAllocation(
+    TensorView* tv) {
+  NVF_ERROR(tv->getMemoryType() == MemoryType::Tensor);
+  // Get all compute-at IDs
+  const auto& loop_domain = tv->getLoopDomain();
+  std::unordered_set<IterDomain*> ca_ids(
+      loop_domain.begin(), loop_domain.begin() + tv->getComputeAtPosition());
+  // Filter allocation domain
+  std::vector<IterDomain*> lane;
+  std::vector<IterDomain*> column;
+  const auto& raw_allocation_domain = tv->getMaybeAllocationDomain();
+  const int64_t dimsep = tv->getTMemDimSepPos();
+  for (int64_t i : c10::irange((int64_t)raw_allocation_domain.size())) {
+    std::vector<IterDomain*>& target = i < dimsep ? lane : column;
+    IterDomain* id = raw_allocation_domain[i];
+    ParallelType p_type = id->getParallelType();
+    if (ir_utils::isMemorySharedAcross(MemoryType::Tensor, p_type)) {
+      target.push_back(id);
+      continue;
+    }
+    if (id->isBroadcast() || id->isReduction() || id->extent()->isOneInt() ||
+        ir_utils::isMemoryPartitionedAcross(MemoryType::Tensor, p_type) ||
+        ca_ids.count(id)) {
```

Comment:
I suppose we are assuming `raw_allocation_domain` is not that different from `loop_domain` as `ca_ids` is a set of loop IDs, while we are traversing through `raw_allocation_domain`. Can you clarify what we assume here between those two domains?

Reply:
I believe this function has the same set of assumption as https://github.com/NVIDIA/Fuser/blob/e5f7c0c131feb39979ed4b34733e899e74dee99a/csrc/id_model/indexing.cpp#L170
which we should refactor in the future.

We never clearly stated the set of assumptions, but in my opinion, they are:
- an ID in the allocation is either fully allocated or not allocated
- if it is not allocated due to compute-at or parallelization, then this ID must also reside in the loop domain